### PR TITLE
[XDR - IR] add dedup mechanism to fetch

### PR DESCRIPTION
**Pull request by jlevypaloalto**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue
](https://jira-dc.paloaltonetworks.com/browse/XSUP-39965)

## Description
Add a de-duplication mechanism to fetch-incidents.

The mechanism works by maintaining an `offset` (using LastRun) that represents the amount of fetched incidents that occurred at the current `last_fetch` (the key `time` in LastRun).
In this manner, whenever an API call is made to fetch new incidents that occurred from the `last_fetch` time, the incidents that have already been fetched are skipped over by the offset and not re-fetched.

This PR also changes the current mechanism that makes a separate API call for each status configured in the parameters, which can cause incidents to be missed, and fetches the incidents always in the same call.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Review by maimorag — COMMENTED** [review]
looks great!
let's schedule a demo

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR_test.py [hunk 1]
```diff
@@ -1580,3 +1581,91 @@ def test_get_xsoar_close_reasons(mocker):
     }
     mocker.patch.object(demisto, 'internalHttpRequest', return_value=mock_response)
     assert get_xsoar_close_reasons() == list(XSOAR_RESOLVED_STATUS_TO_XDR.keys()) + ['CustomReason1', 'CustomReason 2', 'Foo']
+
+
+@freeze_time('1970-01-01 00:00:00.100')
+def test_fetch_incidents_dedup():
+    """
+    Unit test to verify that incidents that occur in the same instant are not not missed or duplicated.
```
Comment (sapirshuker) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR_test.py:1589]: Change the docstring to be more informative - add cases.

**Review by sapirshuker — COMMENTED** [review]
Great, see my comments!

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 2]
```diff
@@ -387,25 +387,31 @@ def get_tenant_info(self):
         return reply.get('reply', {})
 
     def get_multiple_incidents_extra_data(self, exclude_artifacts, incident_id_list=[], gte_creation_time_milliseconds=0,
-                                          status=None, starred=None, starred_incidents_fetch_window=None,
-                                          page_number=0, limit=100):
+                                          statuses=None, starred=None, starred_incidents_fetch_window=None,
```
[the remark this answers is not in the document]
Comment (jlevypaloalto) [reply Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:390]: Will do

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 3]
```diff
@@ -1125,43 +1119,45 @@ def fetch_incidents(client, first_fetch_time, integration_instance, exclude_arti
...
                 'rawJSON': json.dumps(incident_data),
             }
             if demisto.params().get('sync_owners') and incident_data.get('assigned_user_mail'):
-                incident['owner'] = demisto.findUser(email=incident_data.get('assigned_user_mail')).get('username')
+                incident['owner'] = demisto.findUser(email=incident_data['assigned_user_mail']).get('username')
             # Update last run and add incident if the incident is newer than last fetch
             if incident_data.get('creation_time', 0) > last_fetch:
                 last_fetch = incident_data['creation_time']
+                offset = 1
+            elif incident_data.get('creation_time') == last_fetch:
```
[the remark this answers is not in the document]
Comment (jlevypaloalto) [reply Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1140]: they're all integers

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR_test.py (hunk 1 above)
Comment (jlevypaloalto) [reply Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR_test.py:1589]: Will do

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 4]
```diff
@@ -1125,43 +1119,45 @@ def fetch_incidents(client, first_fetch_time, integration_instance, exclude_arti
...
     except Exception as e:
         if "Rate limit exceeded" in str(e):
             demisto.info(f"Cortex XDR - rate limit exceeded, number of non created incidents is: "
-                         f"'{len(non_created_incidents)}'.\n The incidents will be created in the next fetch")
+                         f"{len(non_created_incidents)!r}.\n The incidents will be created in the next fetch")
         else:
             raise
 
+    next_run = {
+        'incidents_from_previous_run': non_created_incidents,
```
Comment (AradCarmi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1154]: Why do we keep a list of non-created incidents? We agreed we need only the time and offset for the pagination.

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 5]
```diff
@@ -434,7 +439,7 @@ def get_multiple_incidents_extra_data(self, exclude_artifacts, incident_id_list=
             })
         if len(filters) > 0:
             request_data['filters'] = filters
-
+        demisto.debug(f'{request_data=}')
```
Comment (maimorag) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:442]:
```suggestion
        demisto.debug(f'Before sending get_multiple_incidents_extra_data request: {request_data=}')
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 6]
```diff
@@ -1068,54 +1073,42 @@ def update_related_alerts(client: Client, args: dict):
...
 
     # maintain a list of non created incidents in a case of a rate limit exception
     non_created_incidents: list = raw_incidents.copy()
-    next_run = {}
     try:
-        count_incidents = 0
-
-        for raw_incident in raw_incidents:
+        incidents = []
+        for raw_incident in raw_incidents[:max_fetch]:
```
Comment (maimorag) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1111]: why `raw_incidents[:max_fetch]`?
there is a limit parameter

**Review by maimorag — COMMENTED** [review]
great work!

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py (hunk 6 above)
Comment (jlevypaloalto) [reply Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py]: removed

**Review by JudahSchwartz — APPROVED** [review]

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 7]
```diff
@@ -387,25 +385,32 @@ def get_tenant_info(self):
...
+                'keyword': 'asc',
+            }
+        }
         filters: List[Any] = []
         if incident_id_list:
-            incident_id_list = argToList(incident_id_list, transform=lambda x: str(x))
+            incident_id_list = argToList(incident_id_list, transform=str)
             filters.append({"field": "incident_id_list", "operator": "in", "value": incident_id_list})
-        if status:
+        if statuses:
```
Comment (JudahSchwartz) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:409]: why did we need to switch to support a list here?

**Pull request merged**